### ISSUE TITLE
Unable to build using DDS v2.4

### ISSUE THREAD

**Issue by ktf** [FairRootGroup/FairMQ#164]
If I try to build FairMQ with today DDS master, it fails with:

```
DEBUG:O2:FairMQ:0: CMake Error at fairmq/plugins/DDS/CMakeLists.txt:10 (add_library):
DEBUG:O2:FairMQ:0:   Target "FairMQPlugin_dds" links to target "DDS::dds-user-defaults" but the
DEBUG:O2:FairMQ:0:   target was not found.  Perhaps a find_package() call is missing for an
DEBUG:O2:FairMQ:0:   IMPORTED target, or an ALIAS target is missing?
```

I am using FairMQ 1.4.2. Which version will support DDS 2.4?

**Comment by dennisklein** [comment]
We can confirm it is a bug, the meaning of the `DDS::dds-user-defaults` has changed. It should be `DDS::dds_user_defaults_lib` now. We will provide a fix, do you need it also for `1.4.2`?

**Comment by ktf** [comment]
Yes, for `1.4.x`, thanks.

**Comment by dennisklein** [comment]
While it is definitely the wrong target name, your issue seems to be even another one. We can successfully build FairMQ 1.4.2 with `-DBUILD_DDS_PLUGIN=ON` against DDS master.

**Comment by dennisklein** [comment]
Turned out, we did not need this dependency, so we removed it. Pls find the changes included in testing release `v1.4.4`.

**Comment by dennisklein** [comment]
I assume this is resolved, reopen otherwise.

**Comment by ktf** [comment]
Actually it was not fixed by this. The actual issue is fixed by #172. Having a message and motivation of why DDS was not found if BUILD_DDS_PLUGIN=ON would have improved the experience.

**Comment by dennisklein** [comment]
> Actually it was not fixed by this.

What are you talking about? You reported an error about a missing target. I removed the usage of this target alltogether and stated already that this error must have a different origin.

> The actual issue is fixed by #172.

#172 is not the slightest related to the issue reported here ...

> Having a message and motivation of why DDS was not found if BUILD_DDS_PLUGIN=ON would have improved the experience.

We use `find_package(DDS REQUIRED)` if `BUILD_DDS_PLUGIN=ON`. This will produce a fatal error if DDS was not found. Clearly, the message you reported here, was not a result of DDS not found., neither was the issue adressed by #172...

**Comment by ktf** [comment]
You sound offended from my followup, apologies if that is the case, I did not want to sound rude.

That said:

`You reported an error about a missing target` is false. As stated at the top the problem is "Unable to build using DDS master" (released 2.4 now, actually). The missing target was merely one of the issues, which once fixed pointed me to the actual underlaying problem which is that the `DDS_FOUND` was `FALSE` because `dds_intercom.h` is not anymore in `include`, but in `include/DDS`. #172 makes sure the new include location is actually taken into account by `FindDSS.cmake`. 

#172 is related and it does fix the issue I reported, in the sense that with #172 applied I can successfully build FairMQ with DSS 2.4.

We build with `-DDDS_ROOT=<path-to-dds-installation>` and `-DBUILD_DDS_PLUGIN=ON` and without it I do not see a fatal error. The actual log is below. As you can see, without #172, while it complains about missing DDS targets, it does not provide a clear indication of what is wrong (`dds_intercom.h` was not found).

```
++ mkdir -p /Users/ktf/work/active/sw/osx_x86-64/FairMQ/master-1
++ case $ARCHITECTURE in
++ [[ ! -n '' ]]
+++ brew --prefix boost
++ BOOST_ROOT=/usr/local/Cellar/boost/1.70.0
++ [[ ! -n '' ]]
+++ brew --prefix zeromq
++ ZEROMQ_ROOT=/usr/local/Cellar/zeromq/4.3.1_1
++ [[ ! -n '' ]]
+++ brew --prefix msgpack
++ MSGPACK_ROOT=/usr/local/Cellar/msgpack/3.2.0
++ cmake /Users/ktf/work/active/sw/SOURCES/FairMQ/master/0 -DCMAKE_CXX_STANDARD=17 -DCMAKE_BUILD_TYPE=RELWITHDEBINFO -DCMAKE_INSTALL_PREFIX=/Users/ktf/work/active/sw/osx_x86-64/FairMQ/master-1 -DGTEST_ROOT=/Users/ktf/work/active/sw/osx_x86-64/googletest/1.8.0-1 -DBOOST_ROOT=/usr/local/Cellar/boost/1.70.0 -DFAIRLOGGER_ROOT=/Users/ktf/work/active/sw/osx_x86-64/FairLogger/v1.2.0-1 -DZEROMQ_ROOT=/usr/local/Cellar/zeromq/4.3.1_1 -DMSGPACK_ROOT=/usr/local/Cellar/msgpack/3.2.0 -DDDS_ROOT=/Users/ktf/work/active/sw/osx_x86-64/DDS/master-1 -DASIOFI_ROOT=/Users/ktf/work/active/sw/osx_x86-64/asiofi/v0.3.1-1 -DOFI_ROOT=/Users/ktf/work/active/sw/osx_x86-64/ofi/v1.7.1-1 /Users/ktf/work/active/sw/osx_x86-64/ofi/v1.7.1-1 -DDISABLE_COLOR=ON -DBUILD_DDS_PLUGIN=ON -DBUILD_NANOMSG_TRANSPORT=OFF -DBUILD_EXAMPLES=ON -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_INSTALL_BINDIR=bin
-- Found Git: /nix/store/mmnscklq847gikpyxnvl4c0cqah8znf3-git-2.21.0/bin/git (found version "2.21.0") 
-- The CXX compiler identification is Clang 7.1.0
-- Check for working CXX compiler: /nix/store/1af9gw0gimxrhzzf5zbhajzs9110ff6p-clang-wrapper-7.1.0/bin/clang++
-- Check for working CXX compiler: /nix/store/1af9gw0gimxrhzzf5zbhajzs9110ff6p-clang-wrapper-7.1.0/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- FairMQ 1.4.4-1-g112bd9c-dirty from Tue Jun 25 22:39:52 2019 +0200
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Found Boost 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/Boost-1.70.0
--   Requested configuration: QUIET REQUIRED COMPONENTS container;program_options;filesystem;date_time;regex
-- Found boost_headers 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_headers-1.70.0
-- Found boost_container 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_container-1.70.0
--   libboost_container-mt-x64.dylib
-- Adding boost_container dependencies: headers
-- Found boost_program_options 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_program_options-1.70.0
--   libboost_program_options-mt-x64.dylib
-- Adding boost_program_options dependencies: headers
-- Found boost_filesystem 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_filesystem-1.70.0
--   libboost_filesystem-mt-x64.dylib
-- Adding boost_filesystem dependencies: headers
-- Found boost_date_time 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_date_time-1.70.0
--   libboost_date_time-mt-x64.dylib
-- Adding boost_date_time dependencies: headers
-- Found boost_regex 1.70.0 at /nix/store/pd1pdx5gkjsmvbcsyxbwdp4xail9p261-boost-1.70.0-dev/lib/cmake/boost_regex-1.70.0
--   libboost_regex-mt-x64.dylib
-- Adding boost_regex dependencies: headers
-- Boost 1.64 found.
-- Found Boost components:
   container;program_options;filesystem;date_time;regex
--   
--   CXX STANDARD       C++17 (>= C++11, change with -DCMAKE_CXX_STANDARD=17)
--   
--   GLOBAL CXX FLAGS  -fPIC -O2 -std=c++17
--   
--   BUILD TYPE         CXX FLAGS
--   Debug              -g -Wshadow -Wall -Wextra
--   Release            -O2 -DNDEBUG
-- * RelWithDebInfo     -O2 -g -Wshadow -Wall -Wextra -DNDEBUG
--   Nightly            -O2 -g -Wshadow -Wall -Wextra
--   Profile            -g3 -Wshadow -Wall -Wextra -fno-inline -ftest-coverage -fprofile-arcs
--   Experimental       -O2 -g -Wshadow -Wall -Wextra -DNDEBUG
--   AdressSan          -O2 -g -Wshadow -Wall -Wextra -fsanitize=address -fno-omit-frame-pointer
--   ThreadSan          -O2 -g -Wshadow -Wall -Wextra -fsanitize=thread
--   
--   (Change the build type with -DCMAKE_BUILD_TYPE=...)
--   
--   DEPENDENCY FOUND     VERSION                   PREFIX
--   Boost                . (>= 1.64)               /
--   FairLogger           1.2.0 (>= 1.2.0)          /Users/ktf/work/active/sw/osx_x86-64/FairLogger/v1.2.0-1
--   GTest                unknown (>= 1.7.0)        /Users/ktf/work/active/sw/osx_x86-64/googletest/1.8.0-1
--   ZeroMQ               4.3.1 (>= 4.1.5)          /nix/store/72ngqkqx55xidn0v73yjyz8mfsbcgmc4-zeromq-4.3.1
--   
--   COMPONENT         BUILT?  INFO
--   fairmq             YES    (default, disable with -DBUILD_FAIRMQ=OFF)
--   tests              YES    (default, disable with -DBUILD_TESTING=OFF)
--   nanomsg_transport   NO    (default, enable with -DBUILD_NANOMSG_TRANSPORT=ON)
--   ofi_transport       NO    EXPERIMENTAL (requires C++14) (default, enable with -DBUILD_OFI_TRANSPORT=ON)
--   dds_plugin         YES    (disable with -DBUILD_DDS_PLUGIN=OFF)
--   pmix_plugin         NO    (default, enable with -DBUILD_PMIX_PLUGIN=ON)
--   examples           YES    (default, disable with -DBUILD_EXAMPLES=OFF)
--   docs                NO    (default, enable with -DBUILD_DOCS=ON)
--   
--   RUN STATIC ANALYSIS  OFF  (default, enable with -DRUN_STATIC_ANALYSIS=ON)
--   
-- Configuring done
CMake Error at fairmq/plugins/DDS/CMakeLists.txt:10 (add_library):
  Target "FairMQPlugin_dds" links to target "DDS::dds_intercom_lib" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:10 (add_library):
  Target "FairMQPlugin_dds" links to target "DDS::dds_protocol_lib" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:16 (add_executable):
  Target "fairmq-dds-command-ui" links to target "DDS::dds_intercom_lib" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:16 (add_executable):
  Target "fairmq-dds-command-ui" links to target "DDS::dds_protocol_lib" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:10 (add_library):
  Target "FairMQPlugin_dds" links to target "DDS::dds_intercom_lib" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:10 (add_library):
  Target "FairMQPlugin_dds" links to target "DDS::dds_protocol_lib" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:16 (add_executable):
  Target "fairmq-dds-command-ui" links to target "DDS::dds_intercom_lib" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


CMake Error at fairmq/plugins/DDS/CMakeLists.txt:16 (add_executable):
  Target "fairmq-dds-command-ui" links to target "DDS::dds_protocol_lib" but
  the target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    ASIOFI_ROOT
    MSGPACK_ROOT
    OFI_ROOT


-- Build files have been written to: /Users/ktf/work/active/sw/BUILD/4b1eed5460b39f4ae1c51ed78319a99c225c6e76/FairMQ
```

**Comment by dennisklein** [comment]
Maybe in the future we shall refer to more stable references than a branch name. At the time we responded to this issue, both @rbx and I were testing building FairMQ 1.4.2 against DDS master at that time on 3 different systems and it worked for us. The issue with the new header location must have been introduced to DDS master after we looked at it. So, it is even another underlying problem.

The logs you show in your last post are only possible if the `REQUIRED` argument to `find_package` is not effective. I looked further and found that this is the case for DDS, this is a regression bug introduced by 0ff8eaf84d1cf7f2368c85c9142941345b894db0 in 1.4.0. We use a wrapper around CMake's `find_packge` called `find_package2`. Since 1.4.0 the `VERSION` argument of `find_package2` is a multi-value arg and will eat `REQUIRED` if written like it was the case with DDS. This is the underlying bug for the issue reported here.

A quick fix
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 63404ba..68d524d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -95,7 +95,7 @@ if(BUILD_FAIRMQ)
 endif()

 if(BUILD_DDS_PLUGIN)
-  find_package2(PRIVATE DDS VERSION 2.2 REQUIRED)
+  find_package2(PRIVATE DDS REQUIRED VERSION 2.2)
 endif()

 if(BUILD_PMIX_PLUGIN)
```

will make `REQUIRED` effective again and lead to

```
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find DDS (missing: DDS_INCLUDE_DIR) (found suitable version
  "2.4", minimum required is "2.2")
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindDDS.cmake:55 (find_package_handle_standard_args)
  cmake/FairMQLib.cmake:327 (find_package)
  CMakeLists.txt:98 (find_package2)
```

without #171 or #172.

To summarize up to now:
* Why DDS 2.4 is not found, is understood and we have solutions in #171 and #171
* The underlying cause for the reported missing target error is a non-effective `REQUIRED` which we will fix asap. 
* Why you could not find DDS two weeks ago is still a mystery, but it must have been a negative `DDS_FOUND`, too.

**Comment by ktf** [comment]
Thanks for looking into it.

**Comment by dennisklein** [comment]
I believe the issue is now fixed, when DDS is found I see:

```
$ cmake -DCMAKE_INSTALL_PREFIX=./install -DFAIRLOGGER_ROOT=/home/dklein/projects/FairLogger/build/install -DBUILD_DDS_PLUGIN=ON -DDDS_ROOT=/home/dklein/projects/DDS/build/install ..
-- Found Git: /usr/bin/git (found version "2.21.0")
-- The CXX compiler identification is GNU 9.1.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- FairMQ 1.4.4-8-g8873ac9 from Wed Jun 26 20:22:18 2019 +0200
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
--
--   CXX STANDARD       C++11 (>= C++11, change with -DCMAKE_CXX_STANDARD=17)
--
--   BUILD TYPE         CXX FLAGS
--   Debug              -g -Wshadow -Wall -Wextra
--   Release            -O2 -DNDEBUG
-- * RelWithDebInfo     -O2 -g -Wshadow -Wall -Wextra -DNDEBUG
--   Nightly            -O2 -g -Wshadow -Wall -Wextra
--   Profile            -g3 -Wshadow -Wall -Wextra -fno-inline -ftest-coverage -fprofile-arcs
--   Experimental       -O2 -g -Wshadow -Wall -Wextra -DNDEBUG
--   AdressSan          -O2 -g -Wshadow -Wall -Wextra -fsanitize=address -fno-omit-frame-pointer
--   ThreadSan          -O2 -g -Wshadow -Wall -Wextra -fsanitize=thread
--
--   (Change the build type with -DCMAKE_BUILD_TYPE=...)
--
--   DEPENDENCY FOUND     VERSION                   PREFIX
--   Boost                1.69 (>= 1.67)            /usr
--   DDS                  2.4 (>= 2.4)              /home/dklein/projects/DDS/build/install
--   FairLogger           1.3.0 (>= 1.2.0)          /home/dklein/projects/FairLogger/build/install
--   GTest                unknown (>= 1.7.0)        /usr
--   ZeroMQ               4.3.1 (>= 4.1.5)          /usr
--
--   COMPONENT         BUILT?  INFO
--   fairmq             YES    (default, disable with -DBUILD_FAIRMQ=OFF)
--   tests              YES    (default, disable with -DBUILD_TESTING=OFF)
--   nanomsg_transport   NO    (default, enable with -DBUILD_NANOMSG_TRANSPORT=ON)
--   ofi_transport       NO    EXPERIMENTAL (requires C++14) (default, enable with -DBUILD_OFI_TRANSPORT=ON)
--   dds_plugin         YES    (disable with -DBUILD_DDS_PLUGIN=OFF)
--   pmix_plugin         NO    (default, enable with -DBUILD_PMIX_PLUGIN=ON)
--   examples           YES    (default, disable with -DBUILD_EXAMPLES=OFF)
--   docs                NO    (default, enable with -DBUILD_DOCS=ON)
--
--   RUN STATIC ANALYSIS  OFF  (default, enable with -DRUN_STATIC_ANALYSIS=ON)
--
-- Configuring done
-- Generating done
-- Build files have been written to: /home/dklein/projects/FairMQ2/build
```

When DDS is not found I see:

```
$ cmake -DCMAKE_INSTALL_PREFIX=./install -DFAIRLOGGER_ROOT=/home/dklein/projects/FairLogger/build/install -DBUILD_DDS_PLUGIN=ON -DDDS_ROOT=/home/dklein/projects/DDS/build/install2 ..
-- Found Git: /usr/bin/git (found version "2.21.0")
-- The CXX compiler identification is GNU 9.1.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- FairMQ 1.4.4-8-g8873ac9 from Wed Jun 26 20:22:18 2019 +0200
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
CMake Error at cmake/FairMQLib.cmake:372 (find_package):
  By not providing "FindDDS.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "DDS", but
  CMake did not find one.

  Could not find a package configuration file provided by "DDS" (requested
  version 2.4) with any of the following names:

    DDSConfig.cmake
    dds-config.cmake

  Add the installation prefix of "DDS" to CMAKE_PREFIX_PATH or set "DDS_DIR"
  to a directory containing one of the above files.  If "DDS" provides a
  separate development package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:77 (find_package2)


-- Configuring incomplete, errors occurred!
See also "/home/dklein/projects/FairMQ2/build/CMakeFiles/CMakeOutput.log".
See also "/home/dklein/projects/FairMQ2/build/CMakeFiles/CMakeError.log".
```

This will be available in the testing release `v1.4.5` (for now in `dev`) which will happen once #174 is resolved.
